src/TSTransformer/nodes/class/transformDecorators.ts

```diff
@@ -0,0 +1,289 @@
+			let [expression, prereqs] = state.capture(() => transformExpression(state, decorator.expression));
+
+			luau.list.pushList(result, prereqs);
+
+			if (multipleDecorators) {
+				const tempId = luau.tempId("decorator");
+				luau.list.push(
+					result,
+					luau.create(luau.SyntaxKind.VariableDeclaration, {
+						left: tempId,
+						right: expression,
+					}),
+				);
+				expression = tempId;
+			}
+
+			assert(luau.isIndexableExpression(expression));
+
+			const [key, keyPrereqs] = state.capture(() => transformObjectKey(state, name));
+			luau.list.pushList(result, keyPrereqs);
+
+			// decorator(Class, "name")
+			luau.list.unshift(
+				finalizers,
+				luau.create(luau.SyntaxKind.CallStatement, {
+					expression: luau.call(expression, [classId, key]),
+				}),
+			);
+		}
+	}
+
+	luau.list.pushList(result, finalizers);
+
+	return result;
+}
+
+function transformParameterDecorators(
+	state: TransformState,
+	member: ts.MethodDeclaration | ts.ConstructorDeclaration,
+	classId: luau.AnyIdentifier,
+): luau.List<luau.Statement> {
+	const result = luau.list.make<luau.Statement>();
+
+	for (let i = 0; i < member.parameters.length; i++) {
+		const parameter = member.parameters[i];
+		if (ts.isIdentifier(parameter.name)) {
+			const finalizers = luau.list.make<luau.Statement>();
+			const multipleDecorators = member.decorators !== undefined && member.decorators.length > 1;
+			for (const decorator of parameter.decorators ?? []) {
+				// eslint-disable-next-line no-autofix/prefer-const
+				let [expression, prereqs] = state.capture(() => transformExpression(state, decorator.expression));
+
+				luau.list.pushList(result, prereqs);
+
+				if (multipleDecorators) {
+					const tempId = luau.tempId("decorator");
+					luau.list.push(
+						result,
+						luau.create(luau.SyntaxKind.VariableDeclaration, {
+							left: tempId,
+							right: expression,
+						}),
+					);
+					expression = tempId;
+				}
+
+				assert(luau.isIndexableExpression(expression));
+
+				// decorator(Class, "name", 0)
+				luau.list.unshift(
+					finalizers,
+					luau.create(luau.SyntaxKind.CallStatement, {
+						expression: luau.call(expression, [classId, luau.string(parameter.name.text), luau.number(i)]),
+					}),
+				);
+			}
+			luau.list.pushList(result, finalizers);
+		}
+	}
+
+	return result;
+}
+
+function transformClassDecorators(
+	state: TransformState,
+	node: ts.ClassLikeDeclaration,
+	classId: luau.AnyIdentifier,
+): luau.List<luau.Statement> {
+	const result = luau.list.make<luau.Statement>();
+	const finalizers = luau.list.make<luau.Statement>();
+
+	const multipleDecorators = node.decorators !== undefined && node.decorators.length > 1;
+
+	for (const decorator of node.decorators ?? []) {
+		// eslint-disable-next-line no-autofix/prefer-const
+		let [expression, prereqs] = state.capture(() => transformExpression(state, decorator.expression));
+
+		luau.list.pushList(result, prereqs);
+
+		if (multipleDecorators) {
```

Comment:
Should also include a `&& !luau.isSimple(expression)` to eliminate unnecessary extra tempId assignments

Reply:
Same on lines 26 and 120